src/NServiceBus.Core/Transports/Receiving.cs

```diff
@@ -15,17 +15,24 @@ internal Receiving()
             Defaults(s =>
             {
                 var transportInfrastructure = s.Get<TransportInfrastructure>();
-                var userDiscriminator = s.GetOrDefault<string>("EndpointInstanceDiscriminator");
-                var baseQueueName = s.GetOrDefault<string>("BaseInputQueueName") ?? s.EndpointName();
-                if (userDiscriminator != null)
+                var discriminator = s.GetOrDefault<string>("EndpointInstanceDiscriminator");
+                var baseQueueName = s.GetOrDefault<string>("BaseInputQueueName");
+                var mainInstance = new EndpointInstance(s.EndpointName());
+                if (baseQueueName != null)
                 {
-                    var p = s.Get<TransportInfrastructure>().BindToLocalEndpoint(new EndpointInstance(baseQueueName, userDiscriminator));
-                    s.SetDefault("NServiceBus.EndpointSpecificQueue", transportInfrastructure.ToTransportAddress(new LogicalAddress(p)));
+                    mainInstance = mainInstance.SetProperty("queue", baseQueueName);
```

Comment:
Fair enough. That's a bad stringly-typed API. 

We can limit that to MSMQ but that won't give us the benefits we wanted with regards to samples, at least not all of them.  Let me think about an alternative.